### DISCUSSION TITLE
feature request: Add progress indicator

### DISCUSSION THREAD

**Discussion by kekeimiku** [joaotavora/eglot#835] [Ideas]
I found that lsp-mode works, but I don't want to use lsp-mode.

Would love to have something that indicates the current state, something like this.

![image](https://user-images.githubusercontent.com/36557882/153698365-236cec25-31f1-46a3-9546-434d68aa8f7f.png)

**Comment by apiraino** [comment]
That would be especially useful with the Rust LSP server (`rust-analyzer`, depicted in the screenshot) which on the initial startup takes a relatively long time to index the codebase (~10s or more depending on the number of external dependencies).

Currently I monitor my CPU cores to figure out when `rust-analyzer` has finished its work :)

**Reply by joaotavora** [reply]
If there is an LSP interface for this (I seem to recall there is),  I'll gladly take a pull request that does this.  It doesn't seem terribly hard to do.  Soon, I'll be doing some Rust programming and then I'll probably be using `rust-analyzer`.

**Reply by apiraino** [reply]
(just leaving a note here for everyone)

By looking at how `lsp-mode` implemented [this feature](https://github.com/emacs-lsp/lsp-mode/blob/9957623d93b13fabaca8ba35b85da8fcceaeef69/lsp-mode.el#L6283-L6293), I think the relevant interface should be [`$/progress`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress).

**Reply by nemethf** [reply]
rust-analyzer proposes a couple of protocol extensions.  One of them reports the [server-status](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#server-status).  Maybe that can be used here as well.  I implemented this extension in eglot-x, but I don't think I tried it out on a large project.

**Reply by joaotavora** [reply]
An improved version of the progress indicators requested here are now in the latest version of Eglot, which you can get from GNU-devel ELPA if you want to try it out.

![image](https://user-images.githubusercontent.com/387011/227808968-a32a3531-b867-43e3-8702-ae4733d826a8.png)

**Reply by apiraino** [reply]
@joaotavora I've tried testing this. If I am not mistaken the latest devel version of eglot on GNU-devel ELPA is from March, 16th ([link](https://elpa.gnu.org/devel/eglot.html)), however I see a few more recent patches hinting at changes relevant to this feature request. Example:
- https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=0effdb29d6db4d5d13b09d82db3a681b8501b9d5
- https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=47d8e4b0d3835f5e0b0b6ab76f242ef0aa25bda7

How can I use an external version of eglot? Do I need to recompile Emacs to have the latest version of Eglot? Sorry in advance for the possibly off-topic questions.

has anyone tried and verified if the language server work progress is now visible?

Thanks for your work :)

**Reply by joaotavora** [reply]
> If I am not mistaken the latest devel version of eglot on GNU-devel ELPA is from March, 16th

You are mistaken: it's May, not March

> Do I need to recompile Emacs to have the latest version of Eglot?

No.  What version of Emacs are you on?  If on Emacs 29 pretest, just use the `M-x eglot-upgrade-eglot` command.  If on an earlier version of Emacs, use a package manager. 

"Soon" you'll be able to use `M-x eglot-upgrade-eglot` regardless of the version you're on.
